src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -547,435 +514,354 @@ private ValueGetter<VBuffer<uint>> ComposeGetterVec(IRow input, int iinfo, int s
-            // (value == 0 ? 0 : value) takes care of negative 0, its equal to positive 0 according to the IEEE 754 standard
-            return (Hashing.MixHash(Hashing.MurmurRound(seed, FloatUtils.GetBits(value == 0 ? 0 : value))) & mask) + 1;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint HashCore(uint seed, ref float value, int i, uint mask)
-        {
-            Contracts.Assert(Utils.IsPowerOfTwo(mask + 1));
-            if (float.IsNaN(value))
-                return 0;
-            return (Hashing.MixHash(Hashing.MurmurRound(Hashing.MurmurRound(seed, (uint)i),
-                FloatUtils.GetBits(value == 0 ? 0 : value))) & mask) + 1;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint HashCore(uint seed, ref double value, uint mask)
-        {
-            Contracts.Assert(Utils.IsPowerOfTwo(mask + 1));
-            if (double.IsNaN(value))
-                return 0;
-
-            ulong v = FloatUtils.GetBits(value == 0 ? 0 : value);
-            var hash = Hashing.MurmurRound(seed, Utils.GetLo(v));
-            var hi = Utils.GetHi(v);
-            if (hi != 0)
-                hash = Hashing.MurmurRound(hash, hi);
-            return (Hashing.MixHash(hash) & mask) + 1;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint HashCore(uint seed, ref double value, int i, uint mask)
-        {
-            // If the high word is zero, this should produce the same value as the uint version.
-            Contracts.Assert(Utils.IsPowerOfTwo(mask + 1));
-            if (double.IsNaN(value))
-                return 0;
-
-            ulong v = FloatUtils.GetBits(value == 0 ? 0 : value);
-            var lo = Utils.GetLo(v);
-            var hi = Utils.GetHi(v);
-            var hash = Hashing.MurmurRound(Hashing.MurmurRound(seed, (uint)i), lo);
-            if (hi != 0)
-                hash = Hashing.MurmurRound(hash, hi);
-            return (Hashing.MixHash(hash) & mask) + 1;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint HashCore(uint seed, uint value, uint mask)
-        {
-            Contracts.Assert(Utils.IsPowerOfTwo(mask + 1));
-            if (value == 0)
-                return 0;
-            return (Hashing.MixHash(Hashing.MurmurRound(seed, value)) & mask) + 1;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint HashCore(uint seed, uint value, int i, uint mask)
-        {
-            Contracts.Assert(Utils.IsPowerOfTwo(mask + 1));
-            if (value == 0)
-                return 0;
-            return (Hashing.MixHash(Hashing.MurmurRound(Hashing.MurmurRound(seed, (uint)i), value)) & mask) + 1;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint HashCore(uint seed, ulong value, uint mask)
+        /// <summary>
```

Comment:
It might be worthwhile to move this interface, and its documentation, earlier in the file. I believe it is quite crucial to read this to understand the rest of the code.

Reply:
> The comment is quite important to understand this pull request

I didn't actually write it for this PR, since of course if I just wanted to do that I could have just written it as a comment here, but rather as something whereby I might prevent other bad PRs from being written, where people say, "hey, this looks like something I could easily simplify" to the detriment of our efficiency.

Beyond that though, I would argue it's not only not crucial that it be moved up, but would actually impact the clarity of what is going on. It is not crucial to understand the *rest* of the code in the sense of all the hash transform -- the hash transform code is quite easily understood without this. However it *is* crucial to understanding the next 300 or so lines, specifically. So I'd say if we were to move this interface *and* its comment, it would be, by virtue of being removed from the odd parts of the code that it justifies (you know, those strange parts where I wrote stuff like `THasher hasher = default; hasher.Hash(...);`) it makes the code more clear to have it ***exactly here***, which is, frankly, why I put it here, and not somewhere else.